db/migrations/20190117165648_add_name_to_movies.js

```diff
@@ -0,0 +1,17 @@
+'use strict';
+
+exports.up = async (Knex) => {
+  await Knex.schema.table('movies', (table) => {
+    table.text('name')
+  });
+
+  return Knex.raw('ALTER TABLE movies ALTER COLUMN title DROP NOT NULL');
+};
+
+exports.down = async (Knex) => {
+  await Knex.schema.table('movies', (table) => {
+    table.dropColumn('name')
+  });
+
+  return Knex.raw('ALTER TABLE movies ALTER COLUMN title SET NOT NULL');
```

Comment:
While this query is correct it might have some side effects on big tables. When adding the `NOT NULL` constraint on the `title` column, Postgres will hold an `ACCESS EXCLUSIVE` lock on the table. An `ACCESS EXCLUSIVE` lock, as the name implies, forbids anyone else from interacting with the table until the transaction that holds the lock releases it. This also includes reading from the table (`SELECT ...`) which can cause API outages as it won't be able to return data to customers. A way to avoid holding an `ACCESS EXCLUSIVE` lock is to add the `NOT NULL` constraint but don't validate it. And at a later point (right after) validate it `CONCURRENTLY` which means it won't try to validate the entire table at once. An example of how to accomplish this below:

```js
const TABLE = 'movies';
const CONSTRAINT_NAME = 'title_not_null';

exports.down = async (Knex) => {
  await Knex.schema.table('movies', (table) => {
    table.dropColumn('name')
  });

  await Knex.raw(`ALTER TABLE ${TABLE} ADD CONSTRAINT ${CONSTRAINT_NAME} CHECK(title IS NOT NULL) NOT VALID`);
  return Knex.raw(`ALTER TABLE ${TABLE} VALIDATE CONSTRAINT ${CONSTRAINT_NAME}`);
```

You can read more about locks and conflicting locks [here](https://www.postgresql.org/docs/10/explicit-locking.html).

Also note that for small tables, this doesn't matter as the entire transaction happens in a few milliseconds but for our big tables like `postcards`, `letters`, `address`, etc. this can have a pretty big effect.

Reply:
i didnt know about that, but that makes a lot of sense; thanks for the heads up!